### ISSUE TITLE
mydumper/myloader feature requests

### ISSUE THREAD

**Issue by druud** [maxbube/mydumper#84]
1. mydumper needs a "--delay-innodb-indexes" option,
to make the CREATE TABLE only have the PRIMARY KEY,
and to make myloader add all secondary indexes 
only after the table is fully loaded (like mysqlpump can do).
This needs an extra file, foo.bar-post.sql
And create empty post-files, when "build-empty-files" is active,
both for the "databases/schemas" and for the "tables".
 
2. myloader needs a "-L --log-file" parameter.

**Comment by druud** [comment]
Also nice would be a named config file,
with option groups, similar to my.cnf.

Example for rebuilding a box:
```
[rebuild]
outputdir         %T/tmp
statement-size    4000000
rows              1000000
chunk-filesize    500
regex             '^(?!sys[.])'
build-empty-files
triggers
events
routines
no-locks  # or: less-locking?
# tz-utc  # is the default
logfile    %T/log/mydumper.log
threads    28  # N-4, for the OS
verbose    3
```
Where '%T' would become the 'tmpdir,
see "SELECT @@GLOBAL.datadir, @@GLOBAL.tmpdir;".

Usage example:
mysqldump --options=~/mydumper.cnf:rebuild  #  -O

**Comment by druud** [comment]
4. support nibbling.
Currently, --rows appears to use straight partitions of the MIN(id), MAX(id) range.
But for table with sparse ids, or with a composite Primary Key, a "nibbling" approach is better.

Also, with ranges, always use a closed start and an open and, so ">=" and "<".

And enjoy MySQL's support for multi-column comparisons, see:
https://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#operator_greater-than-or-equal

**Comment by druud** [comment]
5. Don't create empty chunk files, if there will be at least one non-empty chunk.

**Comment by druud** [comment]
6. add hooks, that call shell scripts:
--hook-init-after-metadata (early hook)
--hook-database-after-create (for each database, also once after all database-creates)
--hook-table-after-create (for each table, also once after all table-creates) 
--hook-table-before-data (for each table, also once before all table-dumps)
--hook-table-after-data (for each table, also once after all table-dumps)
--hook-table-before-chunk (for each table chunk)
--hook-table-after-chunk (for each table chunk)
which will be called with parameters --version, --name, --state and --json (optionally, for other metadata).

**Comment by druud62** [comment]
7. add an option to have only one thread active per target

target: table, but could also be a table partition

This to achieve inserting in perfect PK-order, which generally speeds things up, as it prevents page splits.

**Comment by maxbube** [comment]
Hi @druud 

1 and 4 I know are needed for some time now, but will take time to implement. 

About 6 can you elaborate some use cases? I think we are mixing things here, this is a backup tool 

And for 7 is not the same as running without chunks? Maybe it could be helpful only for myloader while still mydumper dumps chunks in parallel.

**Comment by druud62** [comment]
Re. 6: such hook-support generally opens up usage that the author didn't think of before.
It could for example be used for something as simple as creating metrics for monitoring.

I have mainly used mydumper/myloader for rebuilding a 3 TB replicant, to change it into one with innodb_page_size=8k, to optimize it for ReduxIO-iSCSI storage. The dump took 2h40m.

I was only using one box, so I had to wait for the dump to finish, but I could have been loading the dump on a second box already. I also changed the tables, from row_format=compressed to row_format=dynamic, and further did some changes to the partitioning.

**Comment by druud62** [comment]
Re. 7: that was meant for myloader only. AFAICS, now it completely loads one table before loading another table. But the mantra of concurrency is "share anything you want on the reading end, but share nothing on the writing end". So don't have multiple threads writing to the same .ibd-file.

**Comment by vicimikec** [comment]
8. MyLoader basic progress indication

When restoring a database server with a few dozen databases each with a few hundred tables it is nearly impossible to gauge how far along in the process the restore is. It would be awesome if there was even the most basic progress indication in the output of myloader. Something like "105 out of 423 tables restored". It does not need to be precise or take into account the size of one table verses another.

**Comment by druud** [comment]
Re. 8: from the total size of the dump, and the current position(s) in the dump, it can derive a percentage. Best have more than one indication, like "tables done" and "percentage done" and "ETA".

**Comment by saveriomiroddi** [comment]
@maxbube hello Max, would you add a "good first issue" label to issues/feature requests that you think would be good for first contribution? it would give newcomers (including me :smile:) an indication of what's useful for the project and/or it's easy to implement.

**Comment by athompso** [comment]
I just realized one large database I dump fairly often could be vastly sped up by having one thread per partition: as of MySQL 5.6, `SELECT * FROM table PARTITION pname` is valid syntax and avoids the need to reverse-engineer the WHERE clause from the Partition specification.
Would require a separate INFORMATION_SCHEMA lookup to get the list of partition names for each table.

**Comment by davidducos** [comment]
re. 1: https://github.com/maxbube/mydumper/pull/286

**Comment by davidducos** [comment]
I will be closing this issue, as there are already another ticket for each request.
